### PR TITLE
10-2 implement MerkleTreeMap

### PR DESCRIPTION
**Pull request by odeac**

### CODE REVIEW

File: layer2/test.output [hunk 1]
```diff
@@ -0,0 +1,119 @@
+qcheck random seed: 640678644
```
[the remark this answers is not in the document]
Comment (odeac) [reply layer2/test.output:1]: that was committed by mistake

File: layer2/lib/mtm/debug.ml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+let print (msg : string) (a : 'a) : unit =
```
[the remark this answers is not in the document]
Comment (odeac) [reply layer2/lib/mtm/debug.ml:1]: Not really but I need it for debugging so would only remove it just before the PR is merged

File: layer2/chusai.opam [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+authors: ["ovidiu@marigold.dev"]
+homepage: "https://github.com/marigold-dev/chusai"
+bug-reports: "https://github.com/marigold-dev/chusai/issues"
+depends: [
+  "alcotest" {with-test}
+  "qcheck-alcotest" {with-test}
+  "dune" {>= "2.9" & build & >= "2.9"}
+  "cryptokit" {>= "1.17"}
+  "containers" {>= "3.7"}
+  "batteries" {>= "3.5.1"}
```
[the remark this answers is not in the document]
Comment (odeac) [reply layer2/chusai.opam:16]: I'll remove batteries just before the merge, after I don't need the Debug.print anymore.

Also maybe we should have a chusai_tools library where we could put a function like Debug.print. For me it was extremely useful during debugging.

File: layer2/lib/mtm/merkle_tree_map.ml [hunk 4]
```diff
@@ -0,0 +1,670 @@
...
+          }
+      in
+      let n = H.node { current_node.content with left = new_left_node } in
+      op, p, Some n, result
+    in
+    update_opt_node root
+  ;;
+
+  (* A convenience function that wraps a process_tree handler that doesn't return anything and only updates the current node *)
+  let write_only_handler
```
Comment (pecornilleau) [inline layer2/lib/mtm/merkle_tree_map.ml:523]: This is not really an handler, I feel like a name like `make_write_only_handler` would be clearer

**Review by pecornilleau — COMMENTED** [review]
I still want to review the verification in more details, but I'm curious about the `NoChange` and need to prepare for my flight, so I'm publishing what I have for now :)

**Review by pecornilleau — COMMENTED** [review]
I've no major concerns after looking at the proof.

File: layer2/lib/mtm/sha3_256.ml [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+module Make (Serializer : Serializer_algo.SERIALIZER) : Hash_algo.HASH = struct
+  type t = Sha256 of string
+
+  let pp (formatter : formatter) (Sha256 s : t) : unit = Format.fprintf formatter "%s" s
+  let equal (Sha256 left) (Sha256 right) = left = right
+
+  let hash (value : 'v) : t =
+    let open Tezos_base58 in
+    let h =
+      Serializer.serialize value |> Hacl_star.Hacl.SHA3_256.hash |> String.of_bytes
```
[the remark this answers is not in the document]
Comment (odeac) [reply layer2/lib/mtm/sha3_256.ml:13]: @xvw  should I find a replacement for String.of_bytes?

File: layer2/lib/mtm/merkle_tree_map.ml (hunk 4 above)
Comment (odeac) [reply layer2/lib/mtm/merkle_tree_map.ml]: no longer valid. the whole function was removed

File: layer2/lib/mtm/optionext.ml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+module OptionExt : sig 
+  val fold_lazy : none:(unit -> 'b) -> some:('a -> 'b) -> 'a option -> 'b
+end = struct 
+  let fold_lazy ~none ~some opt : 'b =
+    match opt with
+    | None -> none ()
+    | Some x -> some x
+  end
```
[the remark this answers is not in the document]
Comment (odeac) [reply layer2/lib/mtm/optionext.ml]: it is used

File: layer2/lib/test/example_tests.ml [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+;;
+
+let test_from_two_elem_list () =
+  check_lists [ "a", 1; "b", 2 ] (Mtm.to_list @@ Mtm.from_list @@ [ "a", 1; "b", 2 ])
+;;
+
+let test_from_three_elem_list () =
+  check_lists
+    [ "a", 1; "b", 2; "c", 3 ]
+    (Mtm.to_list @@ Mtm.from_list @@ [ "b", 2; "a", 1; "c", 3 ])
```
Comment (rueshyna) [inline layer2/lib/test/example_tests.ml:17]: `Mtm.to_list` and `Mtm.from_list` should be inverse function for each other to me.
I'd suggest either way:
- `Mtm.from_list` rename as `Mtm.build` and `Mtm.to_list` as `Mtm.flatten`
-  replace `list` with `set`
-  add some commons to indicate `Mtm.to_list @@ Mtm.from_lis` aren't identical

File: layer2/lib/test/example_tests.ml [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+;;
+
+let test_remove_missing_key () =
+  check_lists
+    ([ "b", 0 ]
+    |> Mtm.from_list
+    |> Mtm.execute (Mtm.Remove { key = "non-existent" })
+    |> Tools.get_tree
+    |> Mtm.to_list)
+    [ "b", 0 ]
```
Comment (rueshyna) [inline layer2/lib/test/example_tests.ml:36]: the arg's order of excepted and actual is different from other test cases.

File: layer2/lib/test/example_tests.ml [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+    @@ Mtm.from_list [ "a", 1; "b", 2; "c", 3 ])
+;;
+
+let test_remove_node_right_left () =
+  let initial_map = Mtm.from_list [ "a", 1; "c", 3; "b", 2 ] in
+  let _, _, final_map = Mtm.execute (Mtm.Remove { key = "a" }) initial_map in
+  check_lists [ "b", 2; "c", 3 ] (Mtm.to_list final_map)
+;;
+
+let test_remove_2 () =
```
Comment (rueshyna) [inline layer2/lib/test/example_tests.ml:69]: how do it get removed?

File: layer2/lib/mtm/merkle_tree_map.ml [hunk 10]
```diff
@@ -0,0 +1,708 @@
...
+
+    let combine
+        (THash lhash : thash)
+        (KHash khash : khash)
+        (VHash vhash : vhash)
+        (THash rhash : thash)
+      =
+      let h = Hash.(lhash ++ khash ++ vhash ++ rhash) in
+      THash h
+    ;;
```
[the remark this answers is not in the document]
Comment (odeac) [reply layer2/lib/mtm/merkle_tree_map.ml:51]: ocamformat adds them automatically once I added the .ocamlformat file

File: layer2/lib/test/example_tests.ml (hunk 9 above)
Comment (rueshyna) [reply layer2/lib/test/example_tests.ml:69]: the other comments are miner. but just this one is a puzzle to me. is it expected?

File: layer2/lib/test/example_tests.ml (hunk 7 above)
Comment (rueshyna) [reply layer2/lib/test/example_tests.ml]: I saw it changed to `Seq`  but the name still `to_list` and `from_list` on the top.

**Review by pecornilleau — APPROVED** [review]
Seems to be a solid first step to me, I think we shouldn't delay merge more.

**Review by xvw — APPROVED** [review]

**Pull request merged**